clippy_lints/src/no_effect_replace.rs

```diff
@@ -0,0 +1,115 @@
+use clippy_utils::diagnostics::span_lint;
+use if_chain::if_chain;
+use rustc_ast::LitKind;
+use rustc_hir::{BorrowKind, Expr, ExprKind, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `replace` statements which have no effect.
+    ///
+    /// ### Why is this bad?
+    /// It's either a mistake or confusing.
+    ///
+    /// ### Example
+    /// ```rust
+    /// "1234".replace("12", "12");
+    /// "1234".replacen("12", "12", 1);
+    /// ```
+    #[clippy::version = "1.62.0"]
+    pub NO_EFFECT_REPLACE,
+    suspicious,
+    "replace with no effect"
+}
+declare_lint_pass!(NoEffectReplace => [NO_EFFECT_REPLACE]);
+
+impl LateLintPass<'_> for NoEffectReplace {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &'_ Expr<'_>) {
+        let (path, args) = match expr.kind {
+            ExprKind::MethodCall(path, args, _) => (path, args),
+            _ => return,
+        };
+
+        if !(args.len() == 3 || args.len() == 4)
+            || !(path.ident.name.as_str() == "replace" || path.ident.name.as_str() == "replacen")
+        {
+            return;
+        }
+
+        if_chain! {
+            if let ExprKind::Lit(spanned) = &args[1].kind;
+            if let Some(param1) = lit_string_value(&spanned.node);
+
+            if let ExprKind::Lit(spanned) = &args[2].kind;
```

Comment:
Look at [SpanlessEq](https://doc.rust-lang.org/stable/nightly-rustc/clippy_utils/struct.SpanlessEq.html) for a simple generalized solution to comparing the two arguments.